### PR TITLE
FEATURE: Introduce `Neos.Fusion:ActionUri` as replacement for `Neos.Fusion:UriBuilder`

### PR DESCRIPTION
**Pull request by mficzel**
The name  `Neos.Fusion:UriBuilder` stands out from other fusion prototypes because it describes what the prototype does
instead of what will be the result. This contradicts the declarative nature of fusion. In addition it is not possible render links from a subrequests to other main-requests without manipulating the controllerContext. A common case for that is fusion-backend modules that want to render links to the content or other modules.

This change adds the prototype `Neos.Fusion:ActionUri` as a 1:1 replacement that implements the same api but allows to configure the `request` the uri is to be built from. This allows to create links from subrequests into other contexts.

`Neos.Fusion:ActionUri`:
- `request`: (ActionRequest, defaults to the the current ``request``) The action request the uri is built from.
- ... all other attributes equal ``Neos.Fusion:UriBuilder``

Examples: 

```
uri = Neos.Fusion:ActionUri {
    package = 'My.Package'
    controller = 'Registration'
    action = 'new'
}
```

This prototype allows to link between different backend modules like the content-module

```
cotentModuleUri = Neos.Fusion:ActionUri {
    request = ${request.mainRequest}
    package="Neos.Neos.Ui"
    controller="Backend"
    action = 'index'
    arguments.node = ${documentNode}
}
```
or the sites-module which uses a subrequest of the `Backend\\Module` controller.
```
siteModuleUri = Neos.Fusion:ActionUri {
    request = ${request.mainRequest}
    action = "index"
    package = "Neos.Neos"
    controller = "Backend\\Module"
    arguments {
        module = 'administration/sites'
        moduleArguments {
            @action = 'edit'
            site = ${site}
        }
    }
}
```

In addition:

1. The prototype `Neos.Fusion:Link.Action` is adjusted to use `Neos.Fusion:ActionUri` instead of `Neos.Fusion:UriBuilder` for the `href` property.
2. Documentation is added for `Neos.Fusion:Link.Action` and `Neos.Fusion:Link.Resource`

**Review instructions**

The change adds a new implementation instead of modifying or extending the UriBuilderImplementation to avoid breakiness as the `UriBuilder` uses the UriBuilder from the controllerContext while `ActionUri` creates a new UriBuilder.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Documentation/References/NeosFusionReference.rst [hunk 1]
```diff
@@ -655,8 +655,124 @@ A helper object to render the head of an HTTP response
...
+
+A special case is generating URIs for links to Neos modules. In this case often the option `useMainRequest` is needed
+when linking to a controller outside of the context of the current subrequest.
+
+Link to the content module::
+
+	uri = Neos.Fusion:ActionUri {
+		request = ${request.mainRequest}
+    package="Neos.Neos.Ui"
+    controller="Backend"
```
Comment (kdambekalns) [inline Neos.Neos/Documentation/References/NeosFusionReference.rst:695]:
```suggestion
        package="Neos.Neos.Ui"
        controller="Backend"
```

**Review by kdambekalns — DISMISSED** [review]
Looks good by reading, just a tiny whitespace nitpick.

File: Neos.Fusion/Resources/Private/Fusion/Root.fusion [hunk 2]
```diff
@@ -129,6 +129,72 @@ prototype(Neos.Fusion:UriBuilder) {
...
+  arguments = Neos.Fusion:DataStructure
+  argumentsToBeExcludedFromQueryString = Neos.Fusion:DataStructure
+
+  @exceptionHandler = 'Neos\\Fusion\\Core\\ExceptionHandlers\\AbsorbingHandler'
+}
+
+# Renders an link tag with a uri pointing to a controller/action
+#
+# Usage:
+# uri = Neos.Fusion:Link {
```
Comment (mhsdesign) [inline Neos.Fusion/Resources/Private/Fusion/Root.fusion:154]:
```suggestion
# uri = Neos.Fusion:ActionLink {
```

**Review by mhsdesign — CHANGES_REQUESTED** [review]
Thx i generally like it.

I have two concerns:

# 1

in this super long thread https://github.com/neos/neos-development-collection/pull/3255 it was discussed by various people to use this interface for tag links:

`<TagWith.SpecialLink  class="bg-red" href.someprop="a"  href.otherprop="b" />`

i dont know if this is 100% the best choice, but i think we need to stick to one way of handling it.

you suggested this pattern:

`<TagWith.SpecialLink  class="bg-red" someprop="a"  otherprop="b" />`


-----

# 2

the existing
`<Neos.Fusion:Link.Action />`

and your 

`<Neos.Fusion:ActionLink />`

sound really similar. I would like to avoid have two so similar names at the same time in the codebase.

**Comment by mficzel** [comment]
@mhsdesign interesting ... did not know about Neos.Fusion:Link.Action and Neos.Fusion:Link.Resource yet. 

... will convert this back to draft until we figure out a way to unconfuse this

**Comment by mficzel** [comment]
@mhsdesign adjusted the pr and instead of adding a new Neos.Fusion:ActionLink i adjusted the Neos.Fusion:Link.Action to use the Neos.Fusion:ActionUri

Since we never really documented that and the default behavior is equivalent i would consider this still non breaky as it will only behave differently in case someone manipulated the ControllerContext.

@kdambekalns i rerequest your review as this is quite an adjustment.

**Review by mhsdesign — CHANGES_REQUESTED** [review]
Thanks for documenting everything so well :blue_heart: 

Small changes ...

File: Neos.Neos/Documentation/References/NeosFusionReference.rst [hunk 3]
```diff
@@ -655,6 +655,64 @@ A helper object to render the head of an HTTP response
...
+:controller: (string) The controller name (e.g. ``'Registration'``)
+:action: (string) The action name (e.g. ``'new'``)
+:arguments: (array) Arguments to the action by named key
+:format: (string) An optional request format (e.g. ``'html'``)
+:section: (string) An optional fragment (hash) for the URI
+:additionalParams: (array) Additional URI query parameters by named key
+:addQueryString: (boolean) Whether to keep the query parameters of the current URI
+:argumentsToBeExcludedFromQueryString: (array) Query parameters to exclude for ``addQueryString``
+:absolute: (boolean) Whether to create an absolute URI
+:useMainRequest: (boolean) If set, the main Request will be used instead of the current one.
```
[the remark this answers is not in the document]
Comment (mficzel) [reply Neos.Neos/Documentation/References/NeosFusionReference.rst]: That useMainRequest was a leftover from the first approaches that i did not spot during the refactoring. Thanks for fixing

**Review by mhsdesign — APPROVED** [review]
By reading ^^

**Review by kitsunet — APPROVED** [review]

**Pull request merged**